File: src/main/java/dkoryakin/aston/demo/app/AccountApplicationService.java
package dkoryakin.aston.demo.app;

import dkoryakin.aston.demo.api.body.request.AccountCreatePostBody;
import dkoryakin.aston.demo.api.body.request.AccountDepositPostBody;
import dkoryakin.aston.demo.api.body.request.AccountTransferPostBody;
import dkoryakin.aston.demo.api.body.request.AccountWithdrawPostBody;
import dkoryakin.aston.demo.app.result.AccountCreationResult;
import dkoryakin.aston.demo.api.body.response.AccountsGetResponseBody;
import dkoryakin.aston.demo.app.result.OperationResult;
import dkoryakin.aston.demo.domain.service.AccountService;
import dkoryakin.aston.demo.domain.Pin;
import dkoryakin.aston.demo.infrastructure.factory.AccountFactory;
import lombok.RequiredArgsConstructor;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class AccountApplicationService {

    private final AccountService accountService;

    private final AccountFactory factory;

    private final ApplicationEventPublisher eventPublisher;

    public AccountCreationResult createAccount(AccountCreatePostBody body) {
        var account = accountService.createAccount(body.getName(), Pin.valueOf(body.getPin()));
        return AccountCreationResult.success(factory.buildViewFromAccount(account));
    }

    public AccountsGetResponseBody getAllAccounts() {
        var views = accountService.getAllAccounts().stream()
                .map(factory::buildViewFromAccount)
                .toList();
        return new AccountsGetResponseBody(views);
    }

    public OperationResult makeDeposit(Long accountId, AccountDepositPostBody body) {
        var result = accountService.makeDeposit(accountId, body.getAmount());
        result.getEvent().ifPresent(eventPublisher::publishEvent);
        return result;
    }


    public OperationResult makeWithdraw(Long accountId, AccountWithdrawPostBody body) {
        var result = accountService.makeWithdraw(accountId, Pin.valueOf(body.getPin()), body.getAmount());
        result.getEvent().ifPresent(eventPublisher::publishEvent);
        return result;
    }

    public OperationResult makeTransfer(Long accountId, AccountTransferPostBody body) {
        var result = accountService.makeTransfer(accountId, body.getAccountId(), Pin.valueOf(body.getPin()), body.getAmount());
        result.getEvent().ifPresent(eventPublisher::publishEvent);
        return result;
    }

}

File: src/main/java/dkoryakin/aston/demo/app/AccountApplicationService.kt
package dkoryakin.aston.demo.app

import dkoryakin.aston.demo.api.body.request.AccountCreatePostBody
import dkoryakin.aston.demo.api.body.request.AccountDepositPostBody
import dkoryakin.aston.demo.api.body.request.AccountTransferPostBody
import dkoryakin.aston.demo.api.body.request.AccountWithdrawPostBody
import dkoryakin.aston.demo.api.body.response.AccountsGetResponseBody
import dkoryakin.aston.demo.app.result.AccountCreationResult
import dkoryakin.aston.demo.app.result.OperationResult
import dkoryakin.aston.demo.domain.Account
import dkoryakin.aston.demo.domain.Pin
import dkoryakin.aston.demo.domain.event.OperationDoneEvent
import dkoryakin.aston.demo.domain.service.AccountService
import dkoryakin.aston.demo.infrastructure.factory.AccountFactory
import lombok.RequiredArgsConstructor
import org.springframework.context.ApplicationEventPublisher
import org.springframework.stereotype.Service

@Service
@RequiredArgsConstructor
class AccountApplicationService(private val accountService: AccountService,
        private val factory: AccountFactory,
        private val eventPublisher: ApplicationEventPublisher) {

    fun createAccount(body: AccountCreatePostBody): AccountCreationResult {
        val account = accountService.createAccount(body.name, Pin.valueOf(body.pin))
        return AccountCreationResult.success(factory.buildViewFromAccount(account))
    }

    val allAccounts: AccountsGetResponseBody
        get() {
            val views = accountService.retrieveAllAccounts().stream()
                .map { account: Account? -> factory.buildViewFromAccount(account) }
                .toList()
            return AccountsGetResponseBody(views)
        }

    fun makeDeposit(accountId: Long?, body: AccountDepositPostBody): OperationResult {
        val result = accountService.makeDeposit(accountId, body.amount)
        result.event.ifPresent { event: OperationDoneEvent? -> eventPublisher.publishEvent(event) }
        return result
    }

    fun makeWithdraw(accountId: Long, body: AccountWithdrawPostBody): OperationResult {
        val result = accountService.makeWithdraw(accountId, Pin.valueOf(body.pin), body.getAmount())
        result.event.ifPresent { event: OperationDoneEvent? -> eventPublisher.publishEvent(event) }
        return result
    }

    fun makeTransfer(accountId: Long?, body: AccountTransferPostBody): OperationResult {
        val result = accountService!!.makeTransfer(accountId, body.accountId, Pin.valueOf(body.pin), body.amount)
        result.event.ifPresent { event: OperationDoneEvent? -> eventPublisher.publishEvent(event) }
        return result
    }
}
